tensorflow/contrib/rnn/kernels/lstm_ops.cc

```diff
@@ -0,0 +1,950 @@
+        o_tensor->matrix<float>(), ci_tensor->matrix<float>(),
+        co_tensor->matrix<float>(), h_tensor->matrix<float>(),
+        states_grad_tensor->matrix<float>(), h_grad_tensor->matrix<float>(),
+        cs_prev_tensor.matrix<float>(), states_c_grad_tensor.matrix<float>(),
+        states_h_grad_tensor.matrix<float>(), xh_tensor->matrix<float>(),
+        xh_grad_tensor.matrix<float>(), x_grad_tensor->matrix<float>(),
+        dh_tensor.matrix<float>(), do_tensor.matrix<float>(),
+        dcs_tensor.matrix<float>(), dci_tensor.matrix<float>(),
+        df_tensor.matrix<float>(), di_tensor.matrix<float>(),
+        dicfo_tensor->matrix<float>(), states_prev_grad_tensor->matrix<float>(),
+        w_grad_tensor->matrix<float>(), b_grad_tensor->vec<float>());
+  }
+
+ protected:
+  int64 cell_size_;
+  bool bprop_dx_;
+  bool parallel_dw_;
+};
+
+REGISTER_KERNEL_BUILDER(Name("LSTMCellBlockGrad")    \
+                            .Device(DEVICE_CPU),
+                        LSTMCellBlockGradOp<CPUDevice, false>);
+
+#if GOOGLE_CUDA
+namespace functor {
+  template <>
+  void LSTMCellBlockBprop<GPUDevice, true>::operator()(
+      OpKernelContext* ctx, perftools::gputools::Stream* stream,
+      const GPUDevice& d, bool bprop_dx, bool parallel_dw,
+      typename TTypes<float>::ConstMatrix x,
+      typename TTypes<float>::ConstMatrix states_prev,
+      typename TTypes<float>::ConstMatrix w, typename TTypes<float>::ConstVec b,
+      typename TTypes<float>::ConstMatrix i,
+      typename TTypes<float>::ConstMatrix cs,
+      typename TTypes<float>::ConstMatrix f,
+      typename TTypes<float>::ConstMatrix o,
+      typename TTypes<float>::ConstMatrix ci,
+      typename TTypes<float>::ConstMatrix co,
+      typename TTypes<float>::ConstMatrix h,
+      typename TTypes<float>::ConstMatrix states_grad,
+      typename TTypes<float>::ConstMatrix h_grad,
+      typename TTypes<float>::Matrix cs_prev,
+      typename TTypes<float>::Matrix states_c_grad,
+      typename TTypes<float>::Matrix states_h_grad,
+      typename TTypes<float>::Matrix xh, typename TTypes<float>::Matrix xh_grad,
+      typename TTypes<float>::Matrix x_grad, typename TTypes<float>::Matrix dh,
+      typename TTypes<float>::Matrix do_, typename TTypes<float>::Matrix dcs,
+      typename TTypes<float>::Matrix dci, typename TTypes<float>::Matrix df,
+      typename TTypes<float>::Matrix di, typename TTypes<float>::Matrix dicfo,
+      typename TTypes<float>::Matrix states_prev_grad,
+      typename TTypes<float>::Matrix w_grad,
+      typename TTypes<float>::Vec b_grad);
+
+  extern template struct LSTMCellBlockBprop<GPUDevice, true>;
+}  // namespace functor
+
+REGISTER_KERNEL_BUILDER(Name("LSTMCellBlockGrad")  \
+                            .Device(DEVICE_GPU),   \
+                        LSTMCellBlockGradOp<GPUDevice, true>);
+#endif  // GOOGLE_CUDA
+
+template <typename Device, bool USE_CUBLAS>
+class LSTMBlockOp : public OpKernel {
```

Comment:
oh i see what you're doing here.  this actually could be made more general by accepting a 3-tensor for inputs and emitting a 3-tensor for outputs and a 2-tensor for final_state.

by accepting a list of length sequence_length, you're not allowing the per-step input lengths to change; so minibatches that are shorter still have to allocate a ton of space.

instead, accept a 3-tensor for x with dimensions `[batch, max_length, depth]` or `[max_length, batch, depth]` (whichever leads to fastest execution).  this is how `dynamic_rnn` works.  Now your code will work well with the PaddingFIFOQueue (and tf.input.batch, etc)